segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessorTest.java

```diff
@@ -985,6 +986,115 @@ public void testGetTableEntries() throws Exception {
         assertTrue(keyVersions.containsAll(getTableEntriesIteratorsResp.getEntries().getEntries().stream().map(e -> e.getKey().getKeyVersion()).collect(Collectors.toList())));
     }
 
+    @Test
+    public void testReadTableEntriesDelta() throws Exception {
+        // Set up PravegaRequestProcessor instance to execute requests against
+        val rnd = new Random(0);
+        String tableSegmentName = "testReadTableEntriesDelta";
+
+        @Cleanup
+        ServiceBuilder serviceBuilder = newInlineExecutionInMemoryBuilder(getBuilderConfig());
+        serviceBuilder.initialize();
+        StreamSegmentStore store = serviceBuilder.createStreamSegmentService();
+        TableStore tableStore = serviceBuilder.createTableStoreService();
+        ServerConnection connection = mock(ServerConnection.class);
+        InOrder order = inOrder(connection);
+        val recorderMock = mock(TableSegmentStatsRecorder.class);
+        PravegaRequestProcessor processor = new PravegaRequestProcessor(store, tableStore, connection, SegmentStatsRecorder.noOp(),
+                recorderMock, new PassingTokenVerifier(), false);
+
+        // Generate keys.
+        ArrayList<HashedArray> keys = generateKeys(3, rnd);
+        HashedArray testValue = generateValue(rnd);
+        TableEntry e1 = TableEntry.unversioned(keys.get(0), testValue);
+        TableEntry e2 = TableEntry.unversioned(keys.get(1), testValue);
+        TableEntry e3 = TableEntry.unversioned(keys.get(2), testValue);
+
+        // Create a table segment and add data.
+        processor.createTableSegment(new WireCommands.CreateTableSegment(1, tableSegmentName, ""));
+        order.verify(connection).send(new WireCommands.SegmentCreated(1, tableSegmentName));
+        verify(recorderMock).createTableSegment(eq(tableSegmentName), any());
+        processor.updateTableEntries(new WireCommands.UpdateTableEntries(2, tableSegmentName, "",
+                getTableEntries(asList(e1, e2, e3)), WireCommands.NULL_TABLE_SEGMENT_OFFSET));
+        verify(recorderMock).updateEntries(eq(tableSegmentName), eq(3), eq(false), any());
+
+        // 1. Now read the table entries where suggestedEntryCount is equal to number of entries in the Table Store.
+        processor.readTableEntriesDelta(new WireCommands.ReadTableEntriesDelta(3, tableSegmentName, "", 0, 3));
+        // Does not work without some delay.
+        TimeUnit.SECONDS.sleep(1);
```

Comment:
Note.

Reply:
This is a classic example of some async operation that hasn't completed yet. Most likely even your 1s delay may not help in some cases if wherever you are running this is too slow. 

What happens here is that somewhere inside the execution of `readTableEntriesDelta` we execute some async operation without waiting for it to finish. However since that's how `PravegaRequestProcessor` works internally, there is little we can do there to help us in this case. To solve this problem we've created this `InlineExecutor` which is a single-thread executor that executes every task "inline" (i.e., awaits its completion before moving on). This is why all the other tests in this class work consistently well.

This means that somewhere in your code (test code or not), you are not using this executor or are doing some other "funny business" (pretty vague term, but it's mostly a "catch-all" scenario). Let's focus on the first case instead (not using `InlineExecutor`).

If we look at the `CompletableFuture` doc (and implementation) we learn the following things:
1. Any continuation callback method (`thenXYZ`) that does not have the `Async` suffix will be executed on the same thread where the previous task completed (and implicitly on the same thread pool).
2. Any continuation callback method that has the `Async` method will execute as follows :
    1. If you pass an Executor as the last argument to that method, the callback code will be transferred to that Executor.
    2. If you do not pass an Executor, then the `ForkJoinPool.commonPool()` will be used instead.

In case 1.2. above, if you have such a callback, your execution will jump out of InlineExecutor to a regular thread pool, thus the runtime will not await its execution to complete so this test will likely not work.

I did a quick check on the code path that this is executing (and compared it to the regular iterator) and found that you do not pass `this.executor` to this method: https://github.com/pravega/pravega/pull/4867/files#diff-3dfd10edcdb2751b3655ff97953c1612R330

There may be others downstream, but I'm willing to bet this is the cause of your issues.